kamon-annotation/src/main/scala/kamon/annotation/instrumentation/BaseAnnotationInstrumentation.scala

```diff
@@ -0,0 +1,180 @@
+    if (method.isAnnotationPresent(classOf[Histogram])) {
+      val histogram = method.getAnnotation(classOf[Histogram])
+      val name = eval(histogram.name())
+      val tags = evalTags(histogram.tags())
+      val dynamicRange = DynamicRange(histogram.lowestDiscernibleValue(), histogram.highestTrackableValue(), histogram.precision())
+      val currentHistogram = Kamon.simpleMetrics.histogram(HistogramKey(name, tags), dynamicRange)
+      histograms.put(methodName(method), currentHistogram)
+    }
+  }
+
+  @inline final def registerMinMaxCounter(method: Method, minMaxCounters: TrieMap[String, MinMaxCounter])(implicit eval: StringEvaluator, evalTags: TagsEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[MinMaxCount])) {
+      val minMaxCount = method.getAnnotation(classOf[MinMaxCount])
+      val name = eval(minMaxCount.name())
+      val tags = evalTags(minMaxCount.tags())
+      val minMaxCounter = Kamon.simpleMetrics.minMaxCounter(MinMaxCounterKey(name, tags))
+      minMaxCounters.put(methodName(method), minMaxCounter)
+    }
+  }
+
+  @inline final def registerCounter(method: Method, counters: TrieMap[String, Counter])(implicit eval: StringEvaluator, evalTags: TagsEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[Count])) {
+      val count = method.getAnnotation(classOf[Count])
+      val name = eval(count.name())
+      val tags = evalTags(count.tags())
+      val counter = Kamon.simpleMetrics.counter(CounterKey(name, tags))
+      counters.put(methodName(method), counter)
+    }
+  }
+
+  @inline final def registerTrace(method: Method, traces: TrieMap[String, TraceContextInfo])(implicit eval: StringEvaluator, evalTags: TagsEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[Trace])) {
+      val count = method.getAnnotation(classOf[Trace])
+      val name = eval(count.value())
+      val tags = evalTags(count.tags())
+      traces.put(methodName(method), TraceContextInfo(name, tags))
+    }
+  }
+
+  @inline final def registerSegment(method: Method, segments: TrieMap[String, SegmentInfo])(implicit eval: StringEvaluator, evalTags: TagsEvaluator): Unit = {
+    if (method.isAnnotationPresent(classOf[Segment])) {
+      val segment = method.getAnnotation(classOf[Segment])
+      val name = eval(segment.name())
+      val category = eval(segment.category())
+      val library = eval(segment.library())
+      val tags = evalTags(segment.tags())
+      segments.put(methodName(method), SegmentInfo(name, category, library, tags))
+    }
+  }
+
+  @inline final def processTrace(traces: TrieMap[String, TraceContextInfo], pjp: ProceedingJoinPoint): AnyRef = {
+    val name = methodName(pjp.getStaticPart.getSignature.asInstanceOf[MethodSignature].getMethod)
+    val traceInfo = traces(name)
+
+    Tracer.withContext(Kamon.tracer.newContext(traceInfo.name)) {
+      traceInfo.tags.foreach { case (key, value) ⇒ Tracer.currentContext.addMetadata(key, value) }
```

Comment:
We need a few changes in the TraceContext for this to work properly. Currently we only have the concept of metadata but the TC should allow you to specify both tags and metadata. The tags are used for the metrics side and should be stable values that will be repeated in many traces, on the other side, metadata should be used to gather specific information about a single trace context and will be of interest for people currently using simple trace subscriptions and when we actually get to distributed traces.

This needs to wait a bit until the actual support for tags gets merged to master.